### PR TITLE
single precision support in skimage.metrics module

### PR DESCRIPTION
**Pull request by grlee77**
## Description

#5219 should be reviewed first since this PR is built on top of that one (to reuse the same `_float_dtype` utility).

Computations of means as in `mse` were intentionally set to use `dtype=float64` and that is preserved here to keep good accuracy on larger images. Similarly, for `structural_similarity`, I added single precision support in most of the function, but kept the final mean computation to generate `mssim` in double precision.

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: skimage/metrics/_structural_similarity.py [hunk 1]
```diff
@@ -226,8 +227,8 @@ def structural_similarity(im1, im2,
     # to avoid edge effects will ignore filter radius strip around edges
     pad = (win_size - 1) // 2
 
-    # compute (weighted) mean of ssim
-    mssim = crop(S, pad).mean()
+    # compute (weighted) mean of ssim. Use float64 for accuracy.
```
Comment (rfezzani) [inline skimage/metrics/_structural_similarity.py:230]: :+1:

File: skimage/metrics/tests/test_structural_similarity.py [hunk 2]
```diff
@@ -77,26 +79,31 @@ def test_structural_similarity_grad(seed):
 
     mssim, grad, s = structural_similarity(
         X, Y, data_range=255, gradient=True, full=True)
+    s.dtype == dtype
+    grad.dtype == dtype
```
Comment (rfezzani) [inline skimage/metrics/tests/test_structural_similarity.py:83]: do you mean
```suggestion
    assert s.dtype == dtype
    assert grad.dtype == dtype
```
?

**Review by rfezzani — APPROVED** [review]
I just left some minor comments, otherwise, everything is fine :wink: thank you @grlee77.

File: skimage/metrics/simple_metrics.py [hunk 3]
```diff
@@ -2,7 +2,7 @@
 from scipy.stats import entropy
 
 from ..util.dtype import dtype_range
-from .._shared.utils import warn, check_shape_equality
+from .._shared.utils import _supported_float_type, warn, check_shape_equality
```
Comment (mkcor) [inline skimage/metrics/simple_metrics.py:5]:
```suggestion
from .._shared.utils import _supported_float_type, check_shape_equality, warn
```

File: skimage/metrics/_structural_similarity.py [hunk 4]
```diff
@@ -1,13 +1,12 @@
 import functools
-from warnings import warn
 
 import numpy as np
 from scipy.ndimage import uniform_filter, gaussian_filter
 
+from .._shared import utils
+from .._shared.utils import check_shape_equality, _supported_float_type, warn
```
Comment (mkcor) [inline skimage/metrics/_structural_similarity.py:7]:
```suggestion
from .._shared.utils import _supported_float_type, check_shape_equality, warn
```

File: skimage/metrics/tests/test_simple_metrics.py [hunk 5]
```diff
@@ -107,14 +113,13 @@ def test_nmi_different_sizes():
     assert normalized_mutual_information(cam[:, :400], cam[:400, :]) > 1
 
 
-def test_nmi_random():
+@pytest.mark.parametrize('dtype', [np.float16, np.float32, np.float64])
+def test_nmi_random(dtype):
     random1 = np.random.random((100, 100))
     random2 = np.random.random((100, 100))
```
Comment (mkcor) [inline skimage/metrics/tests/test_simple_metrics.py:119]:
```suggestion
    random1 = np.random.random((100, 100)).astype(dtype)
    random2 = np.random.random((100, 100)).astype(dtype)
```
Right?

File: skimage/metrics/tests/test_simple_metrics.py [hunk 6]
```diff
@@ -38,10 +39,13 @@ def test_PSNR_vs_IPOL():
     assert_almost_equal(p, p_IPOL, decimal=4)
 
 
-def test_PSNR_float():
+@pytest.mark.parametrize('dtype', [np.float32, np.float64])
```
Comment (mkcor) [inline skimage/metrics/tests/test_simple_metrics.py:42]:
```suggestion
@pytest.mark.parametrize('dtype', [np.float16, np.float32, np.float64])
```
Why not?

**Review by mkcor — COMMENTED** [review]
Thanks, @grlee77: I'm only this far (tiny gesture) from approving/merging!

**Comment by mkcor** [comment]
Weird... at b94bce255adc239131960357a888d23e6d1acdd3 `test_PSNR_float[float16]` is failing because of line 69 in `skimage/metrics/tests/test_simple_metrics.py`; I fetched your branch and saw the same thing locally; to have tests pass, I had to do:
```diff
diff --git a/skimage/metrics/tests/test_simple_metrics.py b/skimage/metrics/tests/test_simple_metrics.py
index a69b9abb3..edcb28fd4 100644
--- a/skimage/metrics/tests/test_simple_metrics.py
+++ b/skimage/metrics/tests/test_simple_metrics.py
@@ -66,7 +66,7 @@ def test_PSNR_float(dtype):
     with expected_warnings(['Inputs have mismatched dtype']):
         p_mixed = peak_signal_noise_ratio(cam / 255.,
                                           np.float32(cam_noisy / 255.))
-    assert_almost_equal(p_mixed, p_float64, decimal=5)
+    assert_almost_equal(p_mixed, p_float64, decimal=decimal)
```
But here on GitHub line 69 *reads* `assert_almost_equal(p_mixed, p_float64, decimal=decimal)` :thinking: 
Rerunning all CI workflows just in case... Have you done any crazy rebasing? :sweat_smile:

**Comment by mkcor** [comment]
Although GitHub would display otherwise, the code base definitely had `assert_almost_equal(p_mixed, p_float64, decimal=5)`, apparently:
decimal
That's why I took the liberty of committing 508415e3affdb259406c65f379e8f7e56d184660, which passed the CI tests.

**Review by mkcor — APPROVED** [review]

**Pull request merged**